### PR TITLE
Parameter with source is card fallback when card/field not found

### PR DESCRIPTION
**Pull request by qnkhuat**
From [product doc](https://www.notion.so/metabase/Custom-value-list-for-dashboard-filters-and-template-tags-parameters-8ddbdfa87b9e4d1c97e77d8b8c16c4ea#7dacee4a8c29464b9a25ecfca01cdd19):
> If the saved question is archived or a value/label column is removed, we revert to the standard list behavior (list of all values, if there are values, plain input box if not).

Epic: https://github.com/metabase/metabase/issues/26898

While trying to deliver the above ask, I combined 2 ns `params.card-values` and `params.static-list` into one single `params.custom-values`. The motivation behind this is:
- the way we have to manually handle each case of `values_source_type` in `api.dashboard, api.card, api.dataset` with each of them having similar flow means we need to have a generic function.
- at first I tried to put the mentioned generic function in `metabase.models.params`, but it got into a nasty cyclic deps. So I decided to combine it with the functionality of the 2 existing ns.

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/28055)

### CODE REVIEW

File: src/metabase/models/params/card_values.clj [hunk 1]
```diff
@@ -1,108 +0,0 @@
-(ns metabase.models.params.card-values
```
Comment (qnkhuat) [inline src/metabase/models/params/card_values.clj:1]: moved to `src/metabase/models/params/custom_values.clj`

File: src/metabase/models/params/static_values.clj [hunk 2]
```diff
@@ -1,28 +0,0 @@
-(ns metabase.models.params.static-values
```
Comment (qnkhuat) [inline src/metabase/models/params/static_values.clj:1]: moved to `src/metabase/models/params/custom_values.clj`

File: src/metabase/models/params/custom_values.clj [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    false
+
+    (not (field->field-info value-field (:result_metadata card)))
+    false
+
+    :else true))
+
+;;; --------------------------------------------- Putting it together ----------------------------------------------
+
+(defn parameter->values
```
Comment (qnkhuat) [inline src/metabase/models/params/custom_values.clj:148]: this and `check-can-get-card-values?` are the only 2 new functions. the rest are just copied over.

File: src/metabase/models/params/custom_values.clj [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    (values-from-card card (:value_field config) query)))
+
+(defn- check-can-get-card-values?
+  [card value-field]
+  (cond
+    (:archived card)
+    false
+
+    (not (field->field-info value-field (:result_metadata card)))
+    false
```
Comment (tsmacdonald) [inline src/metabase/models/params/custom_values.clj:142]: I think this entire fn reduces to

```clojure
(boolean
  (and (not (:archived card))
       (field->field-info value-field (:result_metadata card))))
```

File: src/metabase/models/params/custom_values.clj [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+;;; --------------------------------------------- Putting it together ----------------------------------------------
+
+(defn parameter->values
+  "Given a parameter with a custom-values source, return the values.
+
+  `default-case-fn` is a 0-arity function that returns values list when:
+  - :values_source_type = card but the card is archived or value-field does not exists inside the card.
+  - valeus_source_type = ni."
+  [parameter query default-case-fn]
```
Comment (tsmacdonald) [inline src/metabase/models/params/custom_values.clj:154]: nitpicking but I wonder if the default-case-fn should be a delay instead

File: src/metabase/models/params/custom_values.clj [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+  (case (:values_source_type parameter)
+    "static-list" (static-list-values parameter query)
+    "card"        (let [card (db/select-one Card :id (get-in parameter [:values_source_config :card_id]))]
+                    (when-not (mi/can-read? card)
+                      (throw (ex-info "You don't have permissions to do that." {:status-code 403})))
+                    (if (check-can-get-card-values? card (get-in parameter [:values_source_config :value_field]))
+                      (card-values parameter query)
+                      (default-case-fn)))
```
Comment (tsmacdonald) [inline src/metabase/models/params/custom_values.clj:162]: I wonder if we should log something in this case?

Comment (tsmacdonald) [reply src/metabase/models/params/custom_values.clj:162]: based on the product doc: nevermind, no logging

**Review by tsmacdonald — APPROVED** [review]
This works for me.

One product question I have though is whether, if the card isn't found, the source should be changed. It seems weird to have the UI say the values come from a card (and I guess the card's name would be blank since it doesn't exist??) but then the values actually come from elsewhere.

**Comment by calherries** [comment]
Should we update the `values_source_config` and ``values_source_type` of all parameters when a card is archived, or a field is removed? Rather than having to check all the time?

I guess it's difficult to trigger updating parameters when a field is removed. But it feels like a worthy goal.

File: src/metabase/models/params/custom_values.clj (hunk 5 above)
Comment (qnkhuat) [reply src/metabase/models/params/custom_values.clj:154]: I like passing fn around better. For example: it's easier to update this fn to take an argument like `exception` if we want those fn to have custom behavior based on the context it was called.

**Comment by qnkhuat** [comment]
@tsmacdonald, @calherries both are good points, I'll have a follow-up PR that updates parameter when a card is archived or a field is removed.

**Pull request merged**